src/Xamarin.Android.Build.Tasks/Tasks/GeneratePackageManagerJava.cs

```diff
@@ -269,10 +269,36 @@ void AddEnvironment ()
 				throw new InvalidOperationException ($"Unsupported BoundExceptionType value '{BoundExceptionType}'");
 			}
 
+			int assemblyNameWidth = 0;
 			int assemblyCount = ResolvedAssemblies.Length;
+
+			Action<ITaskItem> updateNameWidth = (ITaskItem assembly) => {
+				string assemblyName = Path.GetFileName (assembly.ItemSpec);
+				if (assemblyName.Length > assemblyNameWidth) {
+					assemblyNameWidth = assemblyName.Length;
```

Comment:
I like the idea, but this isn't "Unicode"-aware: for better or worse, this is entirely valid:

```
% csc /out:😀.exe hw.cs
% mono 😀.exe               
Hello, world!
```

or for the `.csproj`-inclined:

```xml
  <PropertyGroup>
    <AssemblyName>😀</AssemblyName>
  </PropertyGroup>
```

On macOS, what *is* 😀?

```
% ls -l | tail -1 | hexdump -C
00000000  2d 72 77 2d 72 2d 2d 72  2d 2d 20 20 20 20 31 20  |-rw-r--r--    1 |
00000010  6a 6f 6e 20 20 73 74 61  66 66 20 20 20 20 20 20  |jon  staff      |
00000020  33 35 38 34 20 41 75 67  20 31 37 20 31 39 3a 33  |3584 Aug 17 19:3|
00000030  32 20 f0 9f 98 80 2e 65  78 65 0a                 |2 .....exe.|
```

i.e. 😀 is { f0 9f 98 80 }, which is *4* bytes, not 1.

You shouldn't assume that assembly names are restricted to ASCII.

Is this even possible in Xamarin.Android?  I don't know about *emoji*, but we *do* use umlauts within directory names: https://github.com/xamarin/xamarin-android/blob/f7d32102cc3ada67aa12cc0beed2cf97226f41b3/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/AotTests.cs#L159

…which means we should improve that test (and others?) so that *assemblies* also contain umlauts.  Pity we missed that opportunity.

Reply:
Related, in terms of Debug typemap generation: https://github.com/xamarin/java.interop/blob/ff2714200107fb616828b9d1013f69605791d2ba/src/Java.Interop.Tools.JavaCallableWrappers/Java.Interop.Tools.JavaCallableWrappers/TypeNameMapGenerator.cs#L229-L232

Reply:
Good point, I keep forgetting that people can do weird things (especially with emoji) :)